### PR TITLE
relocated AddCharm and AddCharmWithAuth to charms facade and use.

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

This work is in preparation for deploying a CharmHub charm.  AddCharm and AddCharmWithAuthorization facades calls have been updated to both take a charm origin and return one.  This is needed for CharmHub charms to provide an ID found when resolving the charm so it can be used when downloading the charm for deploy or upgrade.  Then passed passed back during the Deploy call and saved.  The ID helps tracking on the charm hub side.

Currently AddCharm and AddCharmWithAuthorization are done via the api client, which then calls a function on the application facade.  Neither of these are versioned.  Deprecate those functions and move the charms facade where they will be versioned.

## QA steps

There should be no change in current functionality deploying local and charm store charms.  Please try each of the following:

1. deploy a cs charm
2. deploy a cs charm at an older revision then upgrade it.
3. deploy a local charm
3. deploy a bundle.

### CODE REVIEW

**Comment by hmlanigan** [comment]
There is no plan to bump the facade version for application, as the functions deprecated are not attached to the client itself.  There is no way to dump the api client facade that I've seen.  The charms facade version was bumped in #11942 .

**Review by SimonRichardson — APPROVED** [review]

**Comment by SimonRichardson** [comment]
$$merge$$

File: apiserver/facades/client/charms/interface.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package charms
+
+import (
+	"github.com/juju/charm/v8"
+	"github.com/juju/errors"
+	"github.com/juju/juju/core/lxdprofile"
```
Comment (manadart) [inline apiserver/facades/client/charms/interface.go:9]: Grouping.

File: cmd/juju/application/deployer/interface.go [hunk 2]
```diff
@@ -22,6 +20,7 @@ import (
 	"github.com/juju/juju/core/model"
 	"github.com/juju/juju/core/series"
 	"github.com/juju/juju/jujuclient"
+	"gopkg.in/macaroon-bakery.v2/httpbakery"
```
Comment (manadart) [inline cmd/juju/application/deployer/interface.go:23]: And here.

File: api/charms/client.go [hunk 3]
```diff
@@ -118,3 +119,49 @@ func (c *Client) CharmInfo(charmURL string) (*CharmInfo, error) {
...
+		Force:  force,
+	}
+	var result params.CharmOriginResult
+	if err := c.facade.FacadeCall("AddCharm", args, &result); err != nil {
+		return apicharm.Origin{}, errors.Trace(err)
+	}
+	return apicharm.APICharmOrigin(result.Origin), nil
+}
+
+// AddCharmWithAuthorization is like AddCharm except it also provides
```
Comment (manadart) [inline api/charms/client.go:144]: :+1:

**Pull request merged**